port/unix/omrsysinfo.c

```diff
@@ -7557,3 +7557,113 @@ omrsysinfo_get_number_context_switches(struct OMRPortLibrary *portLibrary, uint6
 	return OMRPORT_ERROR_SYSINFO_NOT_SUPPORTED;
 #endif /* defined(LINUX) */
 }
+
+/*
+ * Get the process ID and commandline for each process.
+ * @param[in] portLibrary The port library.
+ * @param[in] callback The function to be invoked for each process with the process ID and command info.
+ * @param[in] userData Data passed to the callback.
+ * @return 0 on success, or the first non-zero value returned by the callback.
+ */
+uintptr_t
+omrsysinfo_get_processes(struct OMRPortLibrary *portLibrary, OMRProcessInfoCallback callback, void *userData)
+{
+#if defined(LINUX)
+#define INITIAL_BUFFER_SIZE 4096
+	char path[PATH_MAX];
+	uintptr_t callback_result = 0;
+	uintptr_t pid;
+	intptr_t file = 0;
+	uintptr_t buffer_size = INITIAL_BUFFER_SIZE;
+	intptr_t bytes_read = 0;
+	uintptr_t new_size = 0;
+	intptr_t i = 0;
+	char *end = NULL;
+	char *new_command = NULL;
+	DIR *dir = opendir("/proc");
+	if (NULL == dir) {
+		int32_t rc = findError(errno);
+		portLibrary->error_set_last_error(portLibrary, errno, rc);
+		Trc_PRT_sysinfo_get_processes_failedOpeningProcFS(rc);
+		return (uintptr_t)(intptr_t)rc;
+	}
+	/* Allocate initial buffer. */
+	char *command = (char *)portLibrary->mem_allocate_memory(portLibrary, buffer_size, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+	if (NULL == command) {
+		closedir(dir);
+		return OMRPORT_ERROR_SYSINFO_MEMORY_ALLOC_FAILED;
+	}
+	for (;;) {
+		struct dirent *entry = readdir(dir);
+		if (NULL == entry) {
+			break;
+		}
+		if ('\0' == entry->d_name[0]) {
+			continue;
+		}
+		/* Convert name to pid, skipping non-numeric entries. */
+		pid = (uintptr_t)strtoull(entry->d_name, &end, 10);
+		if ('\0' != *end) {
+			continue;
+		}
+		/* Try reading /proc/[pid]/cmdline. */
+		portLibrary->str_printf(portLibrary, path, sizeof(path), "/proc/%s/cmdline", entry->d_name);
+		file = portLibrary->file_open(portLibrary, path, EsOpenRead, 0);
+		if (0 != file) {
```

Comment:
This test should be `file >= 0` (a negative value signals failure, `0` is a valid file descriptor); same for line 7640.